File: TestParser/TestArr.fs
module TestParser.TestArr

open NUnit.Framework
open CbStyles.Cbon.Parser
open CbStyles.Cbon.Parser.Parser
open CbStyles.Parser
open System.Collections.Generic

[<SetUp>]
let Setup () =
    ()

[<Test>]
let TestArr1 () =
    let code = Reader.reader "[]"
    let r = arr_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ArrToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new List<CbAst>()
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestArr2 () =
    let code = Reader.reader "[ 123 ]"
    let r = arr_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ArrToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new List<CbAst>()
    e.Add(CbAst.Num (ANum "123"))
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestArr3 () =
    let code = Reader.reader "[ asd true ]"
    let r = arr_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ArrToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new List<CbAst>()
    e.Add(CbAst.Str "asd")
    e.Add(CbAst.Bool true)
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestArr4 () =
    let code = Reader.reader "[ asd, true ]"
    let r = arr_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ArrToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new List<CbAst>()
    e.Add(CbAst.Str "asd")
    e.Add(CbAst.Bool true)
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestArr5 () =
    let code = Reader.reader "[ a = 1 ]"
    let f () = arr_loop code |> ignore
    let e = Assert.Throws<ParserError> (new TestDelegate(f))
    printf "%s" (e.Message)

[<Test>]
let TestArr6 () =
    let code = Reader.reader "["
    let r = arr_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ArrToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new List<CbAst>()
    CollectionAssert.AreEqual(e, r)

File: TestParser/TestArr.cs
using CbStyles.Cbon;
using CbStyles.Cbon.Parser;
using NUnit.Framework;
using System.Collections.Generic;

namespace TestParser
{
    public class TestArr
    {
        [SetUp]
        public void Setup() { }

        [Test]
        public void TestArr1()
        {
            var code = "[]";
            Parser.RunInReader(code, code => {
                var r = Parser.ArrLoop(code).Value.Item2;
                Assert.AreEqual(new List<CbVal>(), r);
            });
        }

        [Test]
        public void TestArr2()
        {
            var code = "[ 123 ]";
            Parser.RunInReader(code, code => {
                var r = Parser.ArrLoop(code).Value.Item2;
                Assert.AreEqual(new List<CbVal>() { CbVal.NewNum("123") }, r);
            });
        }

        [Test]
        public void TestArr3()
        {
            var code = "[ asd true ]";
            Parser.RunInReader(code, code => {
                var r = Parser.ArrLoop(code).Value.Item2;
                Assert.AreEqual(new List<CbVal>() { CbVal.NewStr("asd"), CbVal.NewBool(true) }, r);
            });
        }

        [Test]
        public void TestArr4()
        {
            var code = "['asd', null]";
            Parser.RunInReader(code, code => {
                var r = Parser.ArrLoop(code).Value.Item2;
                Assert.AreEqual(new List<CbVal>() { CbVal.NewStr("asd"), CbVal.NewNull() }, r);
            });
        }

        [Test]
        public void TestArr5()
        {
            var code = "[ a = 1 ]";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.ArrLoop(code);
            }));
            Assert.AreEqual("Expected array value but not found \t at 1:5", e.Message);
        }

        [Test]
        public void TestArr6()
        {
            var code = "[";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.ArrLoop(code);
            }));
            Assert.AreEqual("Unexpected EOF \t at 1:1", e.Message);
        }
    }
}
